### PR TITLE
Add HuggingFace Pusher Implementation

### PR DESCRIPTION
**Pull request by deep-diver**
#178 project idea relevant implementation

- [X] Tests pass
- [x] Pre-hook commit pass

### CODE REVIEW

File: tfx_addons/huggingface_pusher/README.md [hunk 1]
```diff
@@ -99,6 +99,11 @@ HFSpaceConfig(
 ## Project Dependencies
 - [tfx](https://pypi.org/project/tfx/)
 - [huggingface-hub](https://pypi.org/project/huggingface-hub/)
+- [Git-LFS](https://git-lfs.github.com/) which could be installed as below
+```shell
+$ curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash
+$ sudo apt-get install git-lfs
+```
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/README.md:106]: why is this needed?

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:106]: @casassg 

since the size of model file is large, HuggingFace Hub leverages Git-LFS feature to support it. So, I left a snippet to let people know that they need to include it in their Dockerfile.

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:106]: so this is not a dependency to use the library itself, but it is a dependency to push things to the remote HuggingFace Hub.

Comment (casassg) [reply tfx_addons/huggingface_pusher/README.md:106]: interesting, is there no way to avoid it from having to be used?

If thats the case, I would recommend on Executor adding a short check that validates wether gitlfs exists and if not fails saying what needs to get added and how (aka point to HF pages)

Guessing HF code will already fail, but may be good to capture the error either before it fails there or capture the exception and parse it correctly for users of the component

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:106]: sounds reasonable! thanks for the advice, will do that

**Comment by deep-diver** [comment]
@casassg 

added a simple function to check git-lfs installed, and added tests for it in the `Executor`

File: tfx_addons/huggingface_pusher/executor.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        Raises:
+            RuntimeError: if app_path is not set when space_config is provided.
+            RuntimeError: if git-lfs is not installed.
+        """
+    self._log_startup(input_dict, output_dict, exec_properties)
+
+    if self._is_git_lfs_installed() is not True:
+      raise RuntimeError(
+          "Git-LFS is not installed. "
+          "Git-LFS installation guide can be found in https://git-lfs.github.com/."
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/executor.py:101]: nitpick: may be better to instead reference HugginFace documentation here instead.

File: tfx_addons/huggingface_pusher/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+It gives the follwing outputs:
+- `pushed` : integer value to denote if the model is pushed or not. This is set to 0 when the input model is not blessed, and it is set to 1 when the model is successfully pushed
+- `pushed_version` : string value to indicate the current model version. This is decided by `time.time()` Python built-in function
+- `repo_id` : model repository ID where the model is pushed to. This follows the format of f"{username}/{repo_name}"
+- `branch` : branch name where the model is pushed to. The branch name is automatically assigned to the same value of  `pushed_version`
+- `commit_id` : the id from the commit history (branch name could be sufficient to retreive a certain version of the model) of the model repository
+- `repo_url` : model repository URL. It is something like f"https://huggingface.co/{repo_id}/{branch}"
+- `space_url` : space repository URL. It is something like f"https://huggingface.co/{repo_id}"
+
+### HuggingFace Space spacific configurations
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/README.md:76]:
```suggestion
### HuggingFace Space specific configurations
```

File: tfx_addons/huggingface_pusher/component_test.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+from tfx_addons.huggingface_pusher.component import HFPusher
+
+
+class HFPusherTest(tf.test.TestCase):
+  def testConstruct(self):
+    test_model = channel_utils.as_channel([standard_artifacts.Model()])
+    hf_pusher = HFPusher(
+        username="test_username",
+        access_token="test_access_token",
+        repo_name="test_repo_name",
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/component_test.py:29]: nit: add space_config example

File: tfx_addons/huggingface_pusher/executor_test.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    mock_runner_function.return_value = {
+        "test_key": "test_value",
+        "repo_url": "test_repo_url"
+    }
+    mock_which.return_value = None
+
+    exe = executor.Executor()
+
+    try:
+      exe.Do({}, {"pushed_model": [standard_artifacts.PushedModel()]}, {})
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/executor_test.py:69]:
```suggestion
      exe.Do({}, {"pushed_model": [standard_artifacts.PushedModel()]}, {})
      assert False
```

File: tfx_addons/huggingface_pusher/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+10. Pushes the commit to the __main__ branch of the HuggingFace Space Repository
+    - only the app on the main branch can be built and run, so the branch for the Space repository should not be managed
+
+
+### Structure
+It takes the following inputs:
+```
+HFModelPusher(
+    username: str,
+    hf_access_token: str,
```
Comment (casassg) [inline tfx_addons/huggingface_pusher/README.md:54]: is this a private token? If so we may want to insert it in a more protected way...

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:54]: Can you guide me how to? I was injecting this token via GitHub Secret in Action

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:54]: or is it enough to add more descriptive information how to pass this token? like GitHub Action.

**Comment by deep-diver** [comment]
> is this a private token? If so we may want to insert it in a more protected way...

except this issue, all comments are addressed

File: tfx_addons/huggingface_pusher/README.md (hunk 6 above)
Comment (casassg) [reply tfx_addons/huggingface_pusher/README.md:54]: There are 2 ways to go about this: 
- Encrypt your token and have a way to decrypt it (see @hanneshapke work -> https://github.com/tensorflow/tfx-addons/tree/main/tfx_addons/message_exit_handler#usage-example-at-digits )
- Using something like GCP SecretManager where you store the Secret and retrieve its value at execution time -> https://cloud.google.com/vertex-ai/docs/pipelines/secret-manager

Comment (deep-diver) [reply tfx_addons/huggingface_pusher/README.md:54]: thank you, so I will write as how `message_exit_handler` achive this goal by optionally letting users to specify `decrypt_fn`.

**Comment by deep-diver** [comment]
@casassg 

addressed your comment. I have re-defined the component spec with `decrypt_fn` to let users optionally specify a function to decrypt `access_token` if the `access_token` is provided in encrypted form.

**Review by casassg — APPROVED** [review]
/lgtm May be to get @sayakpaul to review (btw we need to merge https://github.com/tensorflow/tfx-addons/pull/192 before to get ownership over this folder to you and @sayakpaul )

**Comment by sayakpaul** [comment]
Will do it soon (~day).

**Review by sayakpaul — APPROVED** [review]
